Download custom clang-tidy from s3

1ntEgr8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61121
* **#61120**
* #61119

This change implements downloading the clang-tidy binary from S3. It
does this by lifting the existing implementation of the download routine
(which was in `tools/linter/clang_format_utils.py`) to a generic
`tools/linter/install/download_bin.py`.

Folder organization:
```
tools/linter/install
|_ clang_tidy.py
|_ download_bin.py
|_ hashes/
|_ __init__.py

```

`download_bin.py` contains the code for downloading and verifying a
binary from specified sources.  `clang_tidy.py` declares the inputs for
installing `clang-tidy` `hashes/` contains sha256sum hashes of the
binaries (used for verification)

The command to download a linter binary will look like this:
```
python3 tools/linter/install/<binary-name>.py
```

For example, to install `clang-tidy`:
```
python3 tools/linter/install/clang_tidy.py
```

Based on the inputs declared in `tools/linter/install/clang_tidy.py`,
this will install our custom build of `clang-tidy` at
`$PYTORCH_ROOT/.clang-tidy-bin/clang-tidy`

Test plan:

Successful download:
```
python3 tools/linter/install/clang_tidy.py
.clang-tidy-bin/clang-tidy --checks="*" --list-checks | grep "misc-max-tokens"
```

Unsuccessful download:
```
echo "bad hash" > tools/linter/install/hashes/clang-tidy-linux64
python3 tools/linter/install/clang_tidy.py # this should fail
```

1ntEgr8 committed:
Download custom clang-tidy from s3

This change implements downloading the clang-tidy binary from S3. It
does this by lifting the existing implementation of the download routine
(which was in `tools/linter/clang_format_utils.py`) to a generic
`tools/linter/install/download_bin.py`.

Folder organization:
```
tools/linter/install
|_ clang_tidy.py
|_ download_bin.py
|_ hashes/
|_ __init__.py

```

`download_bin.py` contains the code for downloading and verifying a
binary from specified sources.  `clang_tidy.py` declares the inputs for
installing `clang-tidy` `hashes/` contains sha256sum hashes of the
binaries (used for verification)

The command to download a linter binary will look like this:
```
python3 tools/linter/install/<binary-name>.py
```

For example, to install `clang-tidy`:
```
python3 tools/linter/install/clang_tidy.py
```

Based on the inputs declared in `tools/linter/install/clang_tidy.py`,
this will install our custom build of `clang-tidy` at
`$PYTORCH_ROOT/.clang-tidy-bin/clang-tidy`

Test plan:

Successful download:
```
python3 tools/linter/install/clang_tidy.py
.clang-tidy-bin/clang-tidy --checks="*" --list-checks | grep "misc-max-tokens"
```

Unsuccessful download:
```
echo "bad hash" > tools/linter/install/hashes/clang-tidy-linux64
python3 tools/linter/install/clang_tidy.py # this should fail
```

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Download custom clang-tidy from s3"


This change implements downloading the clang-tidy binary from S3. It
does this by lifting the existing implementation of the download routine
(which was in `tools/linter/clang_format_utils.py`) to a generic
`tools/linter/install/download_bin.py`.

Folder organization:
```
tools/linter/install
|_ clang_tidy.py
|_ download_bin.py
|_ hashes/
|_ __init__.py

```

`download_bin.py` contains the code for downloading and verifying a
binary from specified sources.  `clang_tidy.py` declares the inputs for
installing `clang-tidy` `hashes/` contains sha256sum hashes of the
binaries (used for verification)

The command to download a linter binary will look like this:
```
python3 tools/linter/install/<binary-name>.py
```

For example, to install `clang-tidy`:
```
python3 tools/linter/install/clang_tidy.py
```

Based on the inputs declared in `tools/linter/install/clang_tidy.py`,
this will install our custom build of `clang-tidy` at
`$PYTORCH_ROOT/.clang-tidy-bin/clang-tidy`

Test plan:

Successful download:
```
python3 tools/linter/install/clang_tidy.py
.clang-tidy-bin/clang-tidy --checks="*" --list-checks | grep "misc-max-tokens"
```

Unsuccessful download:
```
echo "bad hash" > tools/linter/install/hashes/clang-tidy-linux64
python3 tools/linter/install/clang_tidy.py # this should fail
```

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Download custom clang-tidy from s3"


This change implements downloading the clang-tidy binary from S3. It
does this by lifting the existing implementation of the download routine
(which was in `tools/linter/clang_format_utils.py`) to a generic
`tools/linter/install/download_bin.py`.

Folder organization:
```
tools/linter/install
|_ clang_tidy.py
|_ download_bin.py
|_ hashes/
|_ __init__.py

```

`download_bin.py` contains the code for downloading and verifying a
binary from specified sources.  `clang_tidy.py` declares the inputs for
installing `clang-tidy` `hashes/` contains sha256sum hashes of the
binaries (used for verification)

The command to download a linter binary will look like this:
```
python3 tools/linter/install/<binary-name>.py
```

For example, to install `clang-tidy`:
```
python3 tools/linter/install/clang_tidy.py
```

Based on the inputs declared in `tools/linter/install/clang_tidy.py`,
this will install our custom build of `clang-tidy` at
`$PYTORCH_ROOT/.clang-tidy-bin/clang-tidy`

Test plan:

Successful download:
```
python3 tools/linter/install/clang_tidy.py
.clang-tidy-bin/clang-tidy --checks="*" --list-checks | grep "misc-max-tokens"
```

Unsuccessful download:
```
echo "bad hash" > tools/linter/install/hashes/clang-tidy-linux64
python3 tools/linter/install/clang_tidy.py # this should fail
```

[ghstack-poisoned]

samestep commented on mypy.ini:
```diff
@@ -39,6 +39,7 @@ files =
     test/test_utils.py,
     tools/linter/clang_format_utils.py,
     tools/linter/clang_tidy,
+    tools/linter/install,
     tools/generate_torch_version.py,
     tools/render_junit.py,
     tools/stats
```
same comment here as https://github.com/pytorch/pytorch/pull/61119#discussion_r662401301

samestep commented on tools/linter/clang_format_utils.py:
```diff
@@ -1,16 +1,5 @@
-
-# PyTorch directory root, derived from the location of this file.
-PYTORCH_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.realpath(__file__))))
+from install.download_bin import download, PYTORCH_ROOT  # type: ignore[import]
```
why is the `# type: ignore[import]` needed here?

samestep commented on tools/linter/install/clang_tidy.py:
```diff
@@ -0,0 +1,18 @@
+import os
+from download_bin import download, PYTORCH_ROOT, HASH_PATH  # type: ignore[import]
```
same question about `# type: ignore[import]` here

1ntEgr8 committed:
Update on "Download custom clang-tidy from s3"


This change implements downloading the clang-tidy binary from S3. It
does this by lifting the existing implementation of the download routine
(which was in `tools/linter/clang_format_utils.py`) to a generic
`tools/linter/install/download_bin.py`.

Folder organization:
```
tools/linter/install
|_ clang_tidy.py
|_ download_bin.py
|_ hashes/
|_ __init__.py

```

`download_bin.py` contains the code for downloading and verifying a
binary from specified sources.  `clang_tidy.py` declares the inputs for
installing `clang-tidy` `hashes/` contains sha256sum hashes of the
binaries (used for verification)

The command to download a linter binary will look like this:
```
python3 tools/linter/install/<binary-name>.py
```

For example, to install `clang-tidy`:
```
python3 tools/linter/install/clang_tidy.py
```

Based on the inputs declared in `tools/linter/install/clang_tidy.py`,
this will install our custom build of `clang-tidy` at
`$PYTORCH_ROOT/.clang-tidy-bin/clang-tidy`

Test plan:

Successful download:
```
python3 tools/linter/install/clang_tidy.py
.clang-tidy-bin/clang-tidy --checks="*" --list-checks | grep "misc-max-tokens"
```

Unsuccessful download:
```
echo "bad hash" > tools/linter/install/hashes/clang-tidy-linux64
python3 tools/linter/install/clang_tidy.py # this should fail
```

[ghstack-poisoned]

janeyx99 commented on tools/linter/install/hashes/clang-tidy-linux64:
```diff
@@ -0,0 +1 @@
+44103f662c866921dabd40d047788935c43402d260be10613ab2111e01726892
```
Is this file supposed to be committed vs in the .gitignore?

samestep replied:
yeah it's similar to these two existing files:

https://github.com/pytorch/pytorch/blob/80cab10534e1a9ee6d2493113f169a2570979651/tools/clang_format_hash/linux64/clang-format-linux64#L1

https://github.com/pytorch/pytorch/blob/80cab10534e1a9ee6d2493113f169a2570979651/tools/clang_format_hash/mac/clang-format-mojave#L1

1ntEgr8 replied:
Yes, we need to commit the hash since its used for verifying the binary.
